Breaking Free from the Client-Server Model (#459)
ipfs/ipfs-blog: src/_blog/ipfs-breaking-free-client-server.md

  - [Trustless Verification and Persistence](#trustless-verification-and-persistence)
  - [Unstoppable Data & Services](#unstoppable-data--services)
  - [Scaling & Inter-Planetary](#scaling--inter-planetary)
- [Conclusion](#conclusion)
- [Links](#links)
  - [Getting Started](#getting-started)
  - [Learning Resources](#learning-resources)
    - [Videos](#videos)
    - [Case Studies](#case-studies)
  - [Other Resources](#other-resources)

## What is the Client-Server Model?

![Only a few companies are responsible for serving up most of the web](../assets/ipfs-breaking-free/1.png)
### Limitations of the Client-Server Model

Today, only a few companies are responsible for serving up most of the web. We have Google, Amazon, Facebook, and Fastly to name a few. These companies combined serve us our ads, web pages, social media, videos, images, and allow us to host servers and services on their infrastructures. To interact with these services, we also have to use their APIs. We send some data to their servers, and wait for a response. We then process that response and do whatever magic we need to do to create our apps and websites.

> There are several problems with this design, from costs to privacy. Free services for example are often funded by ads. Ads can destroy privacy as all our data is routed through a few central points and sold to third parties. Effectively, Free services are paid for by our attention and private information.

Reviewer: ```suggestion
There are several problems with this design, from costs to privacy. Free services for example are often funded by ads. Ads can destroy privacy as all our data is routed through a few central points and sold to third parties. Effectively, free services are paid for by our attention and private information.
```

![The source of truth comes from a central entity, the server](../assets/ipfs-breaking-free/2.png)
### Single Source of Truth

> If you want to view a video, send a message, collaborate with remote teammates, or do anything on the web at all, you’re going through someone’s central server. This principle is fundamental to the client-server model. We use a client, such as a web-browser or chat app, and **communicate with a single entity**, that’s the server in the client-server model. Relying on just one point for our source of truth also opens up a whole host of problems like trust. Today, we have SSL certificates to verify an entity is who they say they are, however that doesn’t do us much good if that service is compromised. Even if we are sure of who the entity is, it still doesn’t help us know how advertisers or others are using our data, or even what data is being collected.

Reviewer: SSL is actually deprecated and replaced by TLS. 

```suggestion
If you want to view a video, send a message, collaborate with remote teammates, or do anything on the web at all, you’re going through someone’s central server. This principle is fundamental to the client-server model. We use a client, such as a web-browser or chat app, and **communicate with a single entity**, that’s the server in the client-server model. Relying on just one point for our source of truth also opens up a whole host of problems like trust. Today, we have TLS certificates to verify an entity is who they say they are, however that doesn’t do us much good if that service is compromised. Even if we are sure of who the entity is, it still doesn’t help us know how advertisers or others are using our data, or even what data is being collected.
```

> Developers building applications often rely on APIs operated by large centralized entities (e.g. Stripe, Google, Cloudinary, Auth0). The more APIs your application relies on, the more fragile it is. If any of the APIs are unavailable, your app stops functioning. Today, a popular way to build is to lean on the APIs of another remote service, which makes the web as a whole very fragile.

Reviewer: ```suggestion
Developers building applications often rely on APIs operated by large centralized entities (e.g. Stripe, Google, Cloudinary, Auth0). The more APIs your application relies on, the more fragile it is. If any of the APIs are unavailable, your app stops functioning.
```
Feels like that sentence repeats what you've already communicated. Was that intentional?

Author: it was intentional, it repeats the same info in different ways to try to get those with less experience really thinking about their API usage. I felt like making this section not hammer the point home could lead to some missing the point, and I feel this one is a very important point.

![Data is addressed by an arbitrary location](../assets/ipfs-breaking-free/3.png)
### Location Based Addressing

> In the client-server model, generally if you want some data, you must know ***where*** it is. If you want some image, you generally need the full URL to that image to retrieve and share it. If that image is no longer available, you either need to search for it, or upload a copy on another server to make it available again (*that's assuming you even have a copy yourself*). After uploading, you must also share the new location of that image. This principle is called **location-based addressing**, and it's how most of the world wide web functions today. You go to a website, it has a URL, and all data from that website is addressed by the location of the data. If the image analogy is too hard to follow, think of how you’d share a video on your favourite video platform; you’d likely just share the URL to the video, also known as the location of the video.

Reviewer: ```suggestion
In the client-server model, if you want some data, you must know ***where*** it is. If you want an image, you typically need the full URL of that image to retrieve and share it. If that image is no longer available, you either need to search for it or upload a copy on another server to make it available again (*that's assuming you even have a copy yourself*). After uploading, you must also share the new location of that image. This principle is called **location-based addressing**, and it's how most of the World Wide Web functions today. You go to a website with a URL, and all data from that website is addressed by the location of the data. If the image analogy is too hard to follow, think of how you’d share a video on your favourite video platform; you’d likely just share the URL to the video, also known as the location of the video.
```

![An outage of just one server can bring several services down](../assets/ipfs-breaking-free/4.png)
### Single Point of Failure

> Think of when [AWS, Fastly, Facebook, or YouTube go offline](https://totaluptime.com/notable-network-and-cloud-outages-of-2021/). These events are treated as news, and often slow or even halt productivity. **When one of these services goes offline, it’s much more impactful than just a single image or video being lost**, you’ve now lost access to your ability to make new data available at all! All the data you have hosted in these central hubs is temporarily completely inaccessible, or worse, permanently. The location of the data cannot be accessed, because it’s offline. You cannot ask for the data from the wider web, and expect to receive it, even if another related entity has the very data you’re looking for.

Reviewer: ```suggestion
Think of when [AWS, Fastly, Facebook, or YouTube go offline](https://totaluptime.com/notable-network-and-cloud-outages-of-2021/). These events are treated as news, and often slow or even halt productivity. **When one of these services goes offline, it’s much more impactful than just a single image or video being lost**, you’ve now lost access to your ability to make new data available at all! All the data you have hosted in these central hubs is completely inaccessible – temporarily or worse, permanently. The location of the data cannot be accessed, because it’s offline. You cannot ask for the data from the wider web and expect to receive it, even if another related entity has the very data you’re looking for.
```

![Companies need to decide what types of content they want to allow/retain on their services](../assets/ipfs-breaking-free/6.png)
### Restricted and <span style="background-color:#333;">Censored</span> Content

Content creators are often completely locked to a few platforms like YouTube, Instagram, and others. Those can be subject to a whole host of rules and regulations as well, far more restrictive than what your local government might allow, and you have no say in when or how those rules change.

If you want to host any content at all on one of these platforms you must ensure you’re abiding by their rules and regulations. This can be quite unfortunate when **popular content gets lost to rules changing or updating**, which is quite common with video and graphic media today. As a content creator sometimes it can feel like an impossibly daunting task to solve this problem. One could setup their own services as mentioned previously, however not everyone knows how to do that effectively, and aren’t prepared for the time and costs associated with doing that.

> We can do better; **breaking free from the client-server model means re-thinking how the web works as we know it today**.  The client-server model represents web2 as we know it today. Love it or  hate it, there’s a clear and strong argument for how the web2  model, referring to the client-server model, can be improved. What do we call this new model?

Reviewer: ```suggestion
We can do better; **breaking free from the client-server model means rethinking how the web works as we know it today**.  The client-server model represents Web2 as we know it today. Love it or hate it, there’s a clear and strong argument for how the Web2 model characterised by the client-server model, can be improved. So what do we call this new model?
```

## The Distributed Web

![The distributed web](../assets/ipfs-breaking-free/8.png)

> This new model is referring to **the distributed web**, so let’s talk about that, and how IPFS fits in. But first, for the uninitiated, let’s briefly discuss what the distributed web is, and what it means to you. Whether you’re a user or a developer, you can benefit from the distributed web, what I think of as web3.

Reviewer: ```suggestion
This new model is **the distributed web**, so let’s talk about that, and how IPFS fits in. But first, for the uninitiated, let’s briefly discuss what the distributed web is, and what it means to you. Whether you’re a user or a developer, you can benefit from the distributed web, which I think of as Web3.
```

In the **distributed model** each user is also providing a piece of the network itself. If a user goes offline, the network functions as normal. If a major node goes down, the network can still function as well by leveraging the local peers. It is the champion of resilience, and the model that will take us forward. Some nodes might be bigger than others, but no single outage can take the entire network down. Let’s take a look at how IPFS fits into all this, and also dive deeper into the distributed model itself.

![IPFS creates mathematically generated fingerprints for your data, called content IDs (CIDs)](../assets/ipfs-breaking-free/10.png)
### Content Addressing

> We went over location-based addressing, now let’s talk about one of the fundamental building blocks of an alternative called **content addressing**. IPFS creates mathematically generated fingerprints for data called content identifiers, or CIDs for short. This step relates to something called IPLD or Inter-Planetary Linked Data which is fundamental to how IPFS works to give us content addressing, **breaking us free from location-based addressing**.

Reviewer: In this paragraph, I'd suggest:
```suggestion
We went over location-based addressing, now let’s talk about one of the fundamental building blocks of an alternative called **content addressing**. IPFS creates mathematically generated fingerprints for data called content identifiers, or CIDs in short. A CID is a deterministic pointer to data because it is derived from the data and is used as an address to that. In other words, CIDs make content addressing possible, thereby **breaking us free from location-based addressing**.
```
- Connecting the dots between CIDs and content addressing
- Avoid introducing IPLD

Author: Why avoid introducing IPLD?

Reviewer: Because It's another big concept that isn't so easy to understand.

![](../assets/ipfs-breaking-free/11.png)
#### The Composition of a CID

Pictured above is the anatomy of a v1 CID, represented in binary.  On the far left of the image, not pictured here would be the multibase prefix. This is actually omitted because when working with binary, there actually isn’t a multibase prefix, you can save that byte. What the multibase prefix does is allows us to know what base encoding was used to create the CID, as IPLD supports many. 

Next we have the version identifier, which is whatever version number we’re working with, in this case, we’re working with a **version 1** CID.

> Now we’re to the multicodec, **dag-pb**, which is indicating this DAG (directed-acyclic graph) is protocol buffer encoded. The multicodec field itself is an unsigned-varint. The list of supported codecs are available in the [github multiformats repository](https://github.com/multiformats/multicodec/blob/master/table.csv).

Reviewer: ```suggestion
Now we have the multicodec, e.g. **dag-pb**, which indicates this DAG (directed-acyclic graph) is protocol buffer encoded. The multicodec field itself is an unsigned-varint. The list of supported codecs is available in the [github multiformats repository](https://github.com/multiformats/multicodec/blob/master/table.csv).
```

> Next up is the multihash, which includes 3 things, a multihash algorithm, a multihash length, and then finally the hash digest itself. You can see here we’re working with a **sha2 hash of 32 bytes in length**, then the hash itself trails off the screen. The [multihash specification](https://github.com/multiformats/multihash) is also available in the multiformats github repository.

Reviewer: ```suggestion
Finally, is the multihash, which can be broken down into three parts: 
- The multihash algorithm
- The multihash length
- The hash digest

You can see here we’re working with a **sha2 hash of 32 bytes in length**, then the hash itself trails off the screen. The [multihash specification](https://github.com/multiformats/multihash) is also available in the multiformats GitHub repository.
```

![You know the data is correct if the CID is valid. CIDs allow you to get data from anywhere and anyone.](../assets/ipfs-breaking-free/12.png)
### Trustless Verification and Persistence

As we went over, **you don’t have to trust who’s sending data to you, as you can run the hash function of the CID** and verify it yourself. IPFS uses CIDs by either looking up the CID in the distributed hash table (the DHT), or by using bitswap and asking their local peers “do you have this CID?”. With this, it no longer matters where the data lives, as we know exactly what we want, so it doesn’t matter who has it, just that *someone* has it. We’ve now broken free from location-based addressing.

> If you want to share a web page, an image, a video, or an article, you know if you send that CID to your friend they can download the exact same version of the data you also saw. As long as at least your node or someone else’s has a copy, you can share it. It wouldn’t matter if the initial hoster removed the data, only that you or someone else has a copy of the data. Through this, **data can live as long as there’s someone who wants it to**. With Filecoin you can even pay storage providers to guarantee your data will live for some time, maybe even forever, but that dive is for another post. In the meantime if you’re looking for a service to persist your data, I recommend checking out [web3.storage](https://web3.storage), [nft.storage](https://nft.storage), and [lighthouse.storage](https://lighthouse.storage). They’re all storage helpers eager to get your data available over IPFS and backed up to Filecoin.

Reviewer: ```suggestion
If you want to share a web page, an image, a video, or an article, you know if you send that CID to your friend they can download the exact same version of the data you also saw. As long as at least your node or someone else’s has a copy, you can share it. It wouldn’t matter if the initial hoster removed the data, only that you or someone else has a copy of the data. Through this, **data can live as long as there’s someone who wants it to**. With [Filecoin](https://filecoin.io/) you can even pay storage providers to cryptographically guarantee your data will live for some time, maybe even forever, but that dive is for another post. 

In the meantime, if you’re looking for a service to persist your data, I recommend checking out [web3.storage](https://web3.storage), [nft.storage](https://nft.storage), and [lighthouse.storage](https://lighthouse.storage). They’re all storage helpers eager to get your data available over IPFS and backed up to Filecoin.
```

> ![In peer-to-peer networks, no single entity or event can cause a service to go down. Data keeps flowing even during an Internet outage or blockage. The network can find new routes around problems.](https://ipfs.io/ipfs/QmSbMnsrf1X95riZ3gZygpkFwoQgXLaMpF6dFibt6QRvrP)

Reviewer: This is a repeat of the image above (11.png). If this is intentional, can you please add this image directly to this repo?

### Unstoppable Data & Services

Now that we’ve broken free of location-based addressing, and moved on to content addressing, what does that mean for us? It means that now transparently, **the network can find new routes around problems**. Problems in this case can be outages, we know that if our node goes down for a period of time that we’re in the clear if we’ve gotten our data onto some other nodes. Or even if a user running an IPFS node happens to have a copy of our data.

> With IPFS your node will cache new data until it’s garbage collected, this  helps with automatically strengthening the resilience of a CID. This can aid with Internet outages as well in certain countries. As long  as some node on the local network has a copy of the data, it can still be served. This shows that in a peer-to-peer network no single  node outage or event can cause the service to go down.

Reviewer: ```suggestion
With IPFS your node will cache new CIDs until they are garbage collected, which helps automatically strengthen the availability of those CIDs. This can aid in regional and even country-wide Internet outages because as long as some node on the local network has a copy of the data, it can still be served. What's more, it shows how in a peer-to-peer network no single node outage can cause the whole service to go down.
```

This knowledge is important when you’re designing your application. You can leverage a gateway, and if you’re tight on time in a hackathon that makes perfect sense. If you want to really create a resilient web3 peer-to-peer application, then it’s important to think about how to achieve resilience. To do that, you must be utilizing the peer-to-peer nature of IPFS. When you rely on a gateway, especially just a single one, then if that gateway slows or goes down, your entire application will go with it. **If you instead use IPFS directly, leveraging it’s peer-to-peer capabilities, then you’re on track to creating a virtually unstoppable service**.

> ![Users re-serve data to their nearby peers automagically. When traffic explodes, the network pitches in to help scale efficiently. Bandwidth scaling is negative](https://ipfs.io/ipfs/QmVk9eCu1dAHpaD5dBJJi2x1ANnp89QtTTBRHKjYdi8K7N)

Reviewer: Seems like another duplicate image. If intentional, can you please add the image directly?

The path is `![](../assets/ipfs-breaking-free/[filename].png)`

### Scaling & Inter-Planetary

We touched on how IPFS users help re-serve data they’re interested in to nearby peers. This happens when a user requests data from the network via a CID, their node caches it, then makes it available to the rest of the network. This aids in more than just resilience, you can see it as the network pitching in to help you scale efficiently too. Think about the client-server problem of when you create a popular app, and your traffic explodes. That situation is effectively flipped on it’s head with IPFS, resulting in negative bandwidth scaling costs.

If a bunch of nodes are attempting to download your CID, they’ll be automatically re-sharing the data as well. While your traffic explodes from people requesting your CID, those very same people are also helping send the data itself to their friends or colleagues automatically. **The more popular your CID is, the easier it is for people to retrieve the data from someone who might be more local to them**.

> This property is very important to the inter-planetary aspect of IPFS. If someone on Mars had some data they originally retrieved from Earth, another Mars user shouldn’t have to wait to retrieve the  same data from Earth. **The network should automatically figure out there’s another node on Mars willing to serve that data**. Content addressing and **the distributed web help us unlock such a future** on the network level, transparently.

Reviewer: ```suggestion
This property is very important to the inter-planetary aspect of IPFS. If someone on Mars had some data they originally retrieved from Earth, another Mars user shouldn’t have to wait to retrieve the same data from Earth. **The network should automatically figure out there’s another node on Mars willing to serve that data**. Content addressing and **the distributed web help us unlock such a future** on the network level, transparently.
```

## Conclusion

> Web3 technologies like IPFS help us exit a world of centralization. This gives more power to the consumer, and less to companies looking to harvest your private information. IPFS also helps us achieve more resilience for our data with a solid distributed design. Below are some choice links for how to begin using IPFS for users just getting started, to developers looking to break into the technology.

Reviewer: ```suggestion
Web3 technologies like IPFS help us exit a world of centralization. This gives more power to the consumer, and less to companies looking to harvest your private information. IPFS also helps us achieve more resilience and interoperability for our data with a solid distributed design. 

If you wanna learn more, below are some links for how to begin using IPFS for new users and for developers looking to delve deeper into the technology.
```
